### PR TITLE
Reimplement argtopk to release the GIL

### LINKED ISSUE #3596: topk vs argtopk Performance
I've run some tests comparing the performance of `topk` vs `argtopk` and it seems like there's a significant difference despite a similar underlying codebase.

Sample script -
```
a = da.random.normal(size=(100000,100),chunks=(1000,100))
b = da.random.normal(size=(100,100000),chunks=(100,10000))

c = a.dot(b)

#top = c.topk(5, axis=-1)
top = c.argtopk(5, axis=-1)

with ProgressBar():
    da.compute(top)
```

When the above is executed with `topk` it completes in 1min 40sec.  When it runs with argtopk, it gets to 4% in 3min 40sec.  I interrupted this test early because a similar experiment took nearly 2 hours to complete - if growth is anywhere near linear, this one would complete in over an hour.

Is this a known issue with argtopk or anything we can do to optimize execution?

### PR DESCRIPTION
**Pull request by crusaderky**
Closes #3596

### CODE REVIEW

File: dask/array/chunk.py [hunk 1]
```diff
@@ -1,248 +1,295 @@
-""" A set of NumPy functions to apply per chunk """
```
Comment (crusaderky) [inline dask/array/chunk.py]: Looks like a github bug getting confused by DOS/Unix line terminations? The commit looks right both in PyCharm and in Gitkraken...

**Comment by crusaderky** [comment]
@piercefreeman @mrocklin ready for review and merge

File: dask/array/chunk.py [hunk 2]
```diff
@@ -169,54 +179,91 @@ def trim(x, axes=None):
...
+
     a = np.partition(a, -k, axis=axis)
-    # return a[-k:] if k>0 else a[:-k], on arbitrary axis
-    return a[[
-        (slice(-k, None) if k > 0 else slice(None, -k))
-        if i == axis else slice(None)
-        for i in range(a.ndim)
-    ]]
+    k_slice = slice(-k, None) if k > 0 else slice(-k)
+    return takeslice(a, k_slice, axis=axis)
```
Comment (mrocklin) [inline dask/array/chunk.py]: I have a slight preference to keep this as it was if it's not critical.  I find that dereferencing small functions like this makes it difficult for casual readers to understand a codebase.  Obviously there is a point where a bit of code is frequently enough used or needs to be tested in isolation that pulling it off into a function makes sense, but it's not clear to me that this has yet gotten to that point.

File: dask/array/chunk.py [hunk 3]
```diff
@@ -169,54 +179,91 @@ def trim(x, axes=None):
...
 
 
-def topk_postprocess(a, k, axis):
-    """Kernel of topk and argtopk.
-    Post-processes the output of topk, sorting the results internally.
+def topk_aggregate(a, k, axis, keepdims):
+    """Final aggregation kernel of topk.
+    Invoke topk one final time and then sort the results internally.
```
Comment (mrocklin) [inline dask/array/chunk.py]: Nitpick, include a space or endline after the first `"""` and separate the header line by an empty line.  See http://numpydoc.readthedocs.io/en/latest/ or http://dask.pydata.org/en/latest/develop.html#docstrings

File: dask/array/reductions.py [hunk 4]
```diff
@@ -38,7 +38,8 @@ def divide(a, b, dtype=None):
 
 
 def reduction(x, chunk, aggregate, axis=None, keepdims=None, dtype=None,
-              split_every=None, combine=None, name=None, out=None):
+              split_every=None, combine=None, name=None, out=None,
+              concatenate=True, output_size=1):
```
Comment (mrocklin) [inline dask/array/reductions.py]: Adding two new keyword arguments to this function seems important.  Can I ask you to explain why they were necessary, perhaps in the top comment of this PR? 

Also, I *wanted* to ask you to also add them to the docstring, but I see that the current docstring is unfortunately quite sprase.  Git blame shows this to be my fault :/

File: dask/array/chunk.py (hunk 2 above)
Comment (crusaderky) [reply dask/array/chunk.py]: I personally find the original incredibly hard to read. Ultimately there should be a numpy PR and a backport to numpy_compat.py. I changed it as you requested for now.

File: dask/array/reductions.py (hunk 4 above)
Comment (crusaderky) [reply dask/array/reductions.py]: I'm writing the whole docstring...

I'd also like to change the function signature to ``reduction(x, chunk, combine, concatenate, ...)`` as I feel the concatenate function is fundamental and should not be relegated to a kwarg. Also it makes it easier for the user to wrap his head around it if the order of the arguments matchhes the order of execution.
Do you agree to the change? I would make it in a separate PR to keep things clean.

Comment (crusaderky) [reply dask/array/reductions.py]: edited typos

File: dask/array/reductions.py [hunk 5]
```diff
@@ -37,11 +37,92 @@ def divide(a, b, dtype=None):
...
+        two steps, one ``chunk ``kernel per input chunk and a single
+        ``aggregate`` kernel at the end. If set to less than that, an
+        intermediate ``combine`` kernel will be used, so that any one
+        ``combine`` or ``aggregate`` kernel has no more than ``split_every``
+        inputs. The depth of the aggregation graph will be
+        :math:`log_{split_every}(input chunks along reduced axes)`. Setting to
+        a low value can reduce cache size and network transfers, at the cost of
+        more CPU and a larger dask graph.
+
+        Omit to let dask heuristically decide a good default. A default can
```
Comment (crusaderky) [inline dask/array/reductions.py]: A _very poor_ heuristic by the way - this will be the object of another PR.

File: dask/array/reductions.py [hunk 6]
```diff
@@ -82,26 +165,28 @@ def _tree_reduce(x, aggregate, axis, keepdims, dtype, split_every=None,
         n = builtins.max(int(split_every ** (1 / (len(axis) or 1))), 2)
         split_every = dict.fromkeys(axis, n)
     else:
-        split_every = dict((k, v) for (k, v) in enumerate(x.numblocks) if k in axis)
+        raise ValueError("split_every must be a int or a dict")
```
Comment (crusaderky) [inline dask/array/reductions.py]: Removed undocumented _feature_ where ``split_every="your mom"`` meant a reduction in exactly 2 passes. We could formally reintroduce it in the form of ``split_every=-1`` as part of a separate PR.

Comment (crusaderky) [reply dask/array/reductions.py]: Also, split_every=None defaults to ``config.get('split_every', 4)``, while split_every=dict() defaults to 2 for the missing axes. Does not seem very coherent to me - but changing it would be out of scope of this PR.

**Comment by crusaderky** [comment]
@mrocklin overhauled all docstrings. Added extensive comments in both topk and argtopk that explain how the algorithm is implemented.

**Comment by crusaderky** [comment]
Travis failure is unrelated and caused by #3578

**Comment by jakirkham** [comment]
Could you please try merging with or rebasing on `master`? Expect that will fix this the issue.

**Comment by piercefreeman** [comment]
Branch seems to take care of my issue in #3596.  LGTM @crusaderky

File: dask/array/tests/test_reductions.py [hunk 7]
```diff
@@ -459,12 +459,6 @@ def test_topk_argtopk1(npfunc, daskfunc, split_every):
     with pytest.raises(ValueError):
         daskfunc(b, -k, axis=3, split_every=split_every)
 
-    # structured arrays
-    assert_eq(npfunc(c, axis=0)[-1:][::-1],
-              daskfunc(c, 1, split_every=split_every))
-    assert_eq(npfunc(c, axis=0)[:1],
-              daskfunc(c, -1, split_every=split_every))
```
Comment (mrocklin) [inline dask/array/tests/test_reductions.py]: Why were these tests removed?  Were they no longer important for some reason?

File: docs/source/changelog.rst [hunk 8]
```diff
@@ -7,7 +7,7 @@ Changelog
 Array
 +++++
 
--
+- Reimplemented ``argtopk`` to make it release the GIL (:pr:`3596`) `Guido Imperiale`
```
Comment (mrocklin) [inline docs/source/changelog.rst]: You'll want to terminate your name here with an underscore like 

    `Guido Imperiale`_

File: dask/array/reductions.py [hunk 9]
```diff
@@ -37,11 +37,92 @@ def divide(a, b, dtype=None):
...
+    output_size: int >= 1, optional
+        Size of the output of the ``aggregate`` kernel along the reduced axes.
+        Ignored if keepdims is False.
+
+    Returns
+    -------
+    dask array
+
+    Kernel Parameters
+    -----------------
```
Comment (mrocklin) [inline dask/array/reductions.py]: My experience has been that sphinx tends to  drop section headers that don't match one of their known set.  I tend to revert to using boldface instead like `**Kernel Parameters**`.   It would be good to check that things haven't changed though.

File: dask/array/reductions.py [hunk 10]
```diff
@@ -37,11 +37,92 @@ def divide(a, b, dtype=None):
...
+              concatenate=True, output_size=1):
     """ General version of reductions
 
-    >>> reduction(my_array, np.sum, np.sum, axis=0, keepdims=False)  # doctest: +SKIP
+    Parameters
+    ----------
+    x: Array
+        Data being reduced along one or more axes
+    chunk: callable(x_chunk, axis, keepdims)
+        First kernel function to be executed when resolving the dask graph.
```
Comment (mrocklin) [inline dask/array/reductions.py]: The term `kernel` might not be well understood by users.  I wonder if we can use the term `function` instead throughout this docstring.  I suspect that this will be more familiar to novice users.

File: dask/array/reductions.py [hunk 11]
```diff
@@ -37,11 +37,92 @@ def divide(a, b, dtype=None):
...
+        If True (the default), the output of the kernels is concatenated into a
+        single np.array before being passed to the ``combine`` and
+        ``aggregate`` kernels. If False, the input of ``combine`` and
+        ``aggregate`` will be either a list of the raw outputs of the previous
+        step or a single output, and the kernel will have to concatenate it
+        itself. It can be useful to set this to False if the chunk and/or
+        combine steps do not produce np.arrays.
+    output_size: int >= 1, optional
+        Size of the output of the ``aggregate`` kernel along the reduced axes.
+        Ignored if keepdims is False.
```
Comment (mrocklin) [inline dask/array/reductions.py]: These seem like good changes to me.  Thank you for adding their explanation.

File: dask/array/tests/test_reductions.py (hunk 7 above)
Comment (crusaderky) [reply dask/array/tests/test_reductions.py]: The previous implementation stuffed the data and the index into a recarray. The test was necessary to verify that you could transparently build a nested recarray. Now that I'm using tuples of arrays, there's no reason to think recarrays will behave any different from anything else.

**Comment by crusaderky** [comment]
@mrocklin on holiday for a week - I'll incorporate your suggestions when I'm back

**Comment by mrocklin** [comment]
Enjoy the time off!

**Comment by crusaderky** [comment]
incorporated all of @mrocklin's suggestions; ready to merge

**Comment by mrocklin** [comment]
Agreed.  Merging.  Thank you for this @crusaderky !
